### PR TITLE
Reconciliation improvements

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area robustness
/kind enhancement
/priority normal

**What this PR does / why we need it**:
This PR adds the `gardener.cloud/operation: reconcile` annotation trigger to ManagedResources. With this annotation it is possible to make the controller reconcile the ManagedResource w/o a spec change.

In addition, a predicate was added to the MR watch in order to trigger a reconciliation whenever the `ResourcesHealthy` condition of a Managed Resource changes from `true --> false`.

**Release note**:

```improvement operator
A forceful reconciliation for managed resources can now be triggered by the annotation `gardener.cloud/operation: reconcile`.
```
```improvement operator
The Gardener-Resource-Manager does now try to recover unhealthy managed resources (condition `ResourcesHealthy`) by reconciling the resource.
```

### CODE REVIEW

**Comment by timuthy** [comment]
I'm a bit concerned about the g/g/extensions dependency. It leads to many more vendor content + the binary has increased by ~7MB. OTOH, we would have code duplication if we implemented the logic in GRM separately. Any thoughts @gardener/gardener-maintainers?

**Comment by tim-ebert** [comment]
> I'm a bit concerned about the g/g/extensions dependency

Yes, me too. I think, g/grm doesn't have a real dependency on g/g as of now, right? (only the hack for the issue/PR templates)

> OTOH, we would have code duplication if we implemented the logic in GRM separately

If the only code we are duplicating here are some straight-forward predicates or similar stuff, I'm completely fine with duplication...
OTOH, if we decouple the `predicate.HasOperationAnnotation` functionality we probably also should use a different annotation like `resources.gardener.cloud/operation` instead of `gardener.cloud/operation`, right?

Long story short, what about just using a predicate to enqueue `ManagedResource`s when the health status changed (just like in the other direction):
```go
extensionspredicate.Or(
	predicate.GenerationChangedPredicate{},
	managerpredicate.ConditionStatusChanged(resourcesv1alpha1.ResourcesHealthy),
),
```

**Comment by timuthy** [comment]
> Long story short, what about just using a predicate to enqueue ManagedResources when the health status changed (just like in the other direction):

Yes, can be one option. What I like about the operation annotation is that it aligns with the approach we already have for other resources in Gardener. It can also be helpful for ops.

**Comment by tim-ebert** [comment]
> What I like about the operation annotation is that it aligns with the approach we already have for other resources in Gardener. It can also be helpful for ops.

Sure, fully agree. It would be really nice to have such a reconcile trigger.

Still, I think using a predicate reacting on the condition for retriggering a reconciliation (just like with a change in `Applied` will trigger a health check) is probably a nicer solution instead of going through an indirection via a specific annotation. WDYT?

**Comment by timuthy** [comment]
> is probably a nicer solution instead of going through an indirection via a specific annotation.

Hm, I don't see the problem why the health controller shouldn't annotate the resource. It's a well defined contract for reconciliation.

**Comment by tim-ebert** [comment]
> Hm, I don't see the problem why the health controller shouldn't annotate the resource. It's a well defined contract for reconciliation.

No, there is no problem. That's fine for me as well.
I just wanted to mention, that for the other direction we only have the `managerpredicate.ConditionStatusChanged` in place and this work pretty well and would also be sufficient for the other usecase. But as said, no problem with achieving it via the annotation.

**Comment by rfranzke** [comment]
Thanks for the PR! A few comments from my side:
1. With respect to your discussion above: I tend to agree with @tim-ebert that GRM could handle this internally with a mapper for the health controller instead of using the detour with the annotation (if there is no real need for it; generally, if we can avoid I/O we should). 
1. You want to GRM reconciliation to always happen when a condition flaps from `True` to some other status. Could this result in too many, potentially non-helpful, reconciliations? I think in our discussions we were only targeting the `ResourcesApplied` condition and only the scenario where it detects that some resources got missing. I'm open to this change, though, what are your thoughts on the "too many reconciliations" question?

**Comment by tim-ebert** [comment]
> 2\. You want to GRM reconciliation to always happen when a condition flaps from `True` to some other status. Could this result in too many, potentially non-helpful, reconciliations? I think in our discussions we were only targeting the `ResourcesApplied` condition and only the scenario where it detects that some resources got missing. I'm open to this change, though, what are your thoughts on the "too many reconciliations" question?

The `ResourcesApplied` condition is only set by the MR controller and doesn't detect, if there is a missing resource. `ResourcesHealthy` OTOH, is set by the health controller and does detect missing resources. I would go for reconciling, if `ResourcesHealthy` changes from `True` to `False` (probably we should avoid reconciliation if the condition flips to `Unknown` or something else, caused by API server not available or whatever...)

**Comment by timuthy** [comment]
Okay, thanks for your feedback. I'll adjust the PR accordingly.

File: pkg/predicate/condition_status.go [hunk 1]
```diff
@@ -22,8 +22,31 @@ import (
...
+	}
+
+	if con2 == nil {
+		return true // condition was removed
+	}
+
+	return con1.Status != con2.Status
+}
+
+// ConditionChangedToUnhealthy compares the given conditions and returns if the `Status` has changed to an unhealthy state.
```
Comment (tim-ebert) [inline pkg/predicate/condition_status.go:42]:
```suggestion
// ConditionChangedToUnhealthy compares the given conditions and returns `true` if the `Status` has changed to an unhealthy state.
```

File: pkg/predicate/condition_status.go [hunk 2]
```diff
@@ -22,8 +22,31 @@ import (
...
+var ConditionChangedToUnhealthy ConditionChangeFn = func(con1, con2 *resourcesv1alpha1.ManagedResourceCondition) bool {
+	return (con1 == nil || con1.Status == resourcesv1alpha1.ConditionTrue) &&
```
Comment (tim-ebert) [inline pkg/predicate/condition_status.go:44]: Can you elaborate, why you want to trigger a reconciliation, if the condition was added (with Status `False`)?
I think, this can't happen anyways... So I guess, this doesn't matter 🤔 
But even if it could happen, an additional reconcile for newly created and broken MRs won't help probably...

Comment (timuthy) [reply pkg/predicate/condition_status.go:44]: Yes, you're right that it's not very likely to happen. Eventually, I wanted it to be on the safe side and thought it wouldn't harm either to reconcile again.

Comment (tim-ebert) [reply pkg/predicate/condition_status.go:44]: You're right. Thanks for clarifying :)
